### ISSUE TITLE
How to analyze memory issues?

### ISSUE THREAD

**Issue by queequac** [AssemblyScript/assemblyscript#1407]
For evaluation purpose I have been porting an existing TypeScript library to AssemblyScript. That was a very lengthy process until the compiler was happy and all constraints have been resolved and I could successfully compile the ported code to JS (using portable std) and WASM at the same time (and all JS tests were green again).

Now I have the issue that the code runs not very well in WASM. I can do very little work, but when repeating the same thing in a loop, I get errors. Or immediatly when calling some of the more complex code paths.

Often I get this:
```
Error: abort:  at ~lib/rt/pure.ts:79:20
    at abort (C:\xxxx\node_modules\@assemblyscript\loader\index.js:73:11)
    at ~lib/rt/pure/__visit (<anonymous>:wasm-function[254]:0xaeb6)
    at ~lib/rt/__visit_members (<anonymous>:wasm-function[266]:0xb2ec)
    at ~lib/rt/pure/markGray (<anonymous>:wasm-function[243]:0xaa80)
    at ~lib/rt/pure/__visit (<anonymous>:wasm-function[254]:0xaec9)
    at ~lib/array/Array<src/node/MyNode>#__visit_impl (<anonymous>:wasm-function[264]:0xb0af)
    at ~lib/rt/__visit_members (<anonymous>:wasm-function[266]:0xb4d7)
    at ~lib/rt/pure/markGray (<anonymous>:wasm-function[243]:0xaa80)
    at ~lib/rt/pure/__visit (<anonymous>:wasm-function[254]:0xaec9)
    at ~lib/rt/__visit_members (<anonymous>:wasm-function[266]:0xb2fd)
```

Already presumably simple changes such as turning
```
const owner: MyNode | null = child.getOwner() ? null : node;
child.setOwner(owner);
```
into `child.setOwner(node);` (makes no sense from a business logic point of view, anyway... btw, getOwner simply returns the private member, while setOwner assigns it) postpones the issue (can do lots more iterations in the loop before crashing)

I found out that setting '--runtime none' removes all my errors... but I guess that is not the solution to my problem, most likely there are several memory leaks waiting to be fixed. 😉 

Unfortunately I have no clue how to find the root cause. I have the feeling I am breaking even more than just one of AssemblyScript's rules without knowing, since it affects so many code paths.

Long story short: How can I debug or analyze such issues? Commenting out various lines to see where the error occurs next seems not to be the most promising approach.

**Comment by MaxGraey** [comment]
you could compile with `--debug --use ASC_RTRACE=1` flags

**Comment by queequac** [comment]
Thanks for your support, @MaxGraey 
Is there more to do? Compiling with the additional --use flag results in 

```
  const instance = new WebAssembly.Instance(module, imports);    
                   ^
TypeError: WebAssembly.Instance(): Import #1 module="rtrace" error: module is not an object or function
```

**Comment by MaxGraey** [comment]
https://github.com/AssemblyScript/assemblyscript/blob/master/lib/rtrace/README.md

**Comment by queequac** [comment]
Mhm, okay... onerror is never called, oninfo shows a lots of allocations, nearly no incremented refCounters and also some frees. 

All I see is ending with
```
....
ALLOC 65040
++ 65040 @ 0->1
ALLOC 65104
++ 65104 @ 0->1
ALLOC 65136
++ 65136 @ 0->1
ALLOC 65168
++ 65168 @ 0->1
ALLOC 65200
++ 65200 @ 0->1
++ 65200 @ 1->2
ALLOC 65232
ALLOC 65264
++ 65264 @ 0->1
++ 65232 @ 0->1
ALLOC 65328
++ 65328 @ 0->1
++ 65328 @ 1->2
ALLOC 65360
++ 65360 @ 0->1
++ 65360 @ 1->2
ALLOC 65392
++ 65392 @ 0->1
++ 65392 @ 1->2
ALLOC 65424
++ 65424 @ 0->1
++ 65424 @ 1->2
ALLOC 65456
++ 65456 @ 0->1
++ 65456 @ 1->2
ALLOC 65488
++ 65488 @ 0->1
++ 65488 @ 1->2
FREE 7984 @ 0
```
before the crash.

Unfortunately I am not really able to relate those allocations to my code. And allocating stuff is not necessarily wrong, is it?

**Comment by dcodeIO** [comment]
Given the information so far provided here, the error is happening at
 https://github.com/AssemblyScript/assemblyscript/blob/5eb99ee028d23c4023956f3bee5f0a6dd3aee23c/std/assembly/rt/pure.ts#L79

during deferred cycle collection. The algorithm in question essentially marks all blocks within the cycle buffer `gray` (possible member of a cycle) before scanning whether these `gray` blocks are in use (`black`) or garbage (`white`). Now what appears to be happening is that a block with RC=0 is being marked gray, which is unexpected, indicating that either a block became corrupted or a block remained in the algorithm for too long. Hard to tell exactly, but it's certainly wrong state somewhere.

Potential causes are unsafe stores into memory in user code corrupting GC headers, a bug in the algorithm or an out of bounds write somewhere in stdlib.

Unfortunately, these problems are notoriously hard to debug, and RTRACE goes into the right direction, yet RTRACE still assumes that certain categories of bugs do not happen, which the one here is one of, so the information it provides might not be sufficient. Perhaps a few questions:

* Are you using any unsafe features, like writing into memory using low-level builtins like `store`, or is your codebase rather idiomatic TS relying solely on stdlib?
* Are you mutating memory externally?
* How does the inheritance hierarchy of your classes look? Classes forming cycles or classes involving virtual overloads are of particular interest given the type of bug seen here.
* Do you see a chance to make a condensed (that's very important) reproducible test case?

**Comment by MaxGraey** [comment]
Also it could relate to wrong manually handling on host side like double call `__release` for already deallocated reference (use after free). So will be great to see js code relate to loader

**Comment by queequac** [comment]
Hi @dcodeIO , let me briefly summarize and answer some of your questions.

> * Are you using any unsafe features, like writing into memory using low-level builtins like `store`, or is your codebase rather idiomatic TS relying solely on stdlib?

No usage of low-level built-ins. This is a port of a typescript library that shall still transpile to JS, so only stdlib where necessary. Compiles also with --noUnsafe flag.

> * Are you mutating memory externally?
> * Are you calling runtime helpers like `release` externally (using the loader)?

To make debugging more easy and so I don't have to deal with the loader until later, I have temporarily put everything into the library itself.
This means in contrast to the JS build output I have a different entry point for AS that exports only one function called `start()` which in turn runs the test code with loops and calculations - in this case already in the WASM. Once everything is working I'd start tinkering around with the loader and these helpers, removing the test code but allowing the host to access my API and its objects directly.

> * How does the inheritance hierarchy of your classes look?

No inheritance at all. There is no extends keyword in the whole code base.

> Classes forming cycles or classes involving virtual overloads are of particular interest given the type of bug seen here.

Have to double-check, at least cycles between classes I cannot rule out for the moment.

> * Do you see a chance to make a condensed (that's very important) reproducible test case?

Unfortunately I cannot make it public right now. And as mentioned before, I do not see code paths thare are not affected and how I could thus easily share a condensed sample without the sharing the whole project. Maybe some further experiments are resulting in separate test cases.

**Comment by MaxGraey** [comment]
Anyway it will be great if you can to isolate the problem to a minimal reproducible example and share here
